enhancements/machine-config/custom-ignition-machineconfig.md

```diff
@@ -0,0 +1,199 @@
+---
+title: Store user ignition customizations in MachineConfig
+authors:
+  - "@hardys"
+  - "@celebdor"
+reviewers:
+  - "@celebdor"
+  - "@cgwalters"
+  - "@crawford"
+  - "@kirankt"
+  - "@runcom"
+  - "@stbenjam"
+  - "@yuqi-zhang"
+approvers:
+  - "@cgwalters"
+  - "@crawford"
+  - "@runcom"
+  - "@yuqi-zhang"
+
+creation-date: 2020-08-21
+last-updated: 2020-11-13
+status: implementable
+---
+
+# Store user ignition customizations in MachineConfig
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes changing the way installer ignition customizations
+are stored, so that instead of storing the modified pointer ignition in a
+Secret we include the user changes in a MachineConfig, such that the MCO
+can manage it, and it is included in the MCO rendered config.
+
+## Motivation
+
+The installer supports user [modification of the pointer ignition config](https://github.com/openshift/installer/blob/master/docs/user/customization.md#os-customization-unvalidated) it generates.  While this interface is marked as unvalidated, we know from [previous bug reports](https://bugzilla.redhat.com/show_bug.cgi?id=1881703) that some users are using it.
```

Comment:
There seems to be a subset of ignition customizations that don't translate directly into machine configs, right? The example in the installer docs shows adjusting some ignition-specific knobs like http timeouts. What do we do about those?

Maybe we should explicitly enumerate the things we can pull out of the stub ignition.

Reply:
> There seems to be a subset of ignition customizations that don't translate directly into machine configs, right? The example in the installer docs shows adjusting some ignition-specific knobs like http timeouts. What do we do about those?

I've been thinking about this very aspect as well. None are pretty. One option is to keep the customized pointer config intact and pass it unchanged along with the newly created machineconfigs and the other option is to start peeling off the customizations(in the pointer config) one by one until we are done.

> Maybe we should explicitly enumerate the things we can pull out of the stub ignition.

I like this idea. I was looking at the ignition struct and there are lots of options to sort through.

Reply:
> There seems to be a subset of ignition customizations that don't translate directly into machine configs,

I'm not sure that's the case but could be wrong :)  I thought the MCO rendered config could contain any ignition to be applied during the initial deployment, but only a subset can be updated subsequently via the MCD?

If that's true then it's probably OK since the installer customizations via ignition only impact the initial deployment (and scale out)?

This shows the supported/unsupported sections for the MCD:

https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigDaemon.md#supported-vs-unsupported-ignition-config-changes

AFAICS from the code the render controller just merges the raw ignition without filtering any sections:

https://github.com/openshift/machine-config-operator/blob/4fbe13c52f4d5cd1f21b12a0bcc3cb3506978953/pkg/controller/common/helpers.go#L71

I guess we can test to confirm by creating a MachineConfig object with one of the MCD unsupported sections and inspect the rendered config.

Reply:
@runcom @yuqi-zhang what are your thoughts on the above - will it be reasonable/supported to encapsulate pointer ignition customizations in a MachineConfig if they e.g contain filesystem/disk changes which only apply for initial deployment?

Reply:
I may be confused, mostly because I've never touched these ignition knobs but the installer shows a customization of the stub ignition to adjust http timeouts for ignition itself:

```
    "timeouts": {
      "httpResponseHeaders": 120
    }
```

How does that become a machineconfig?

Reply:
Right, customizations in the pointer config aren't reflected into the cluster.    This relates to https://github.com/openshift/machine-config-operator/issues/1720 as well as https://github.com/openshift/machine-config-operator/blob/master/docs/FAQ.md#q-if-i-change-something-manually-on-the-host-will-the-mco-revert-it

It would be really nice to capture these and record them in the cluster in the MCO for sure; it's a bit tricky because the node only sees the full rendered config and we'd need to "re-split" it internally.

Reply:
(Sorry, now I understand that solving this issue is what this enhancement is about)

Reply:
> How does that become a machineconfig?

IIUC that ignition can just be encapsulated by a MachineConfig in the installer, and we'll be able to read it combined with the rendered config from the MCO render controller.